### PR TITLE
EDAM tool panel

### PR DESCRIPTION
**Pull request by hexylena**
part of #10587. This was honestly pretty easy and it looks great? (Minus the hundreds of unannotated tools.)

image

<details>
  <summary>Original Notes</summary>
  

I just took the existing ITP loading code and replaced it with edam generated topics rather than whatever section they requested.

Sorting based on EDAM would be nice (edit: done!) E.g. the following image:

</details>

Bugs:

- ~we have a copy of the EDAM ontology in the codebase. Surely there's a better way to handle this?~
- Binary server setting, doesn't permit some of the fancy things people wanted in #10587
- Search works but "show sections" in search is now broken
- ~All the tests fail, even though I made this opt-in?~
- ~Sorting based on EDAM ordering would probably be better.~

### CODE REVIEW

**Comment by hexylena** [comment]
Went ahead an annotated some of my Circos tools and it's fantastic, I love not having to read through the sections to try and figure out where something should go, it's just automatic.

image

**Comment by mvdbeek** [comment]
> * All the tests fail, even though I made this opt-in?

It's just the unit tests that mock out a config object in various places. Want me to fix that ?

File: lib/galaxy/webapps/galaxy/config_schema.yml
Comment (mvdbeek) [inline lib/galaxy/webapps/galaxy/config_schema.yml]:
```suggestion
      beta_edam_toolbox_ontology_path:
        type: str
        default: EDAM_1.23.tsv
        path_resolves_to: data_dir
        required: false
        desc: |
          Sets the path to EDAM ontology file.

      beta_edam_ontology_url:
        type: str
        default: <url_to_wherever we can fetch the ontology>
        required: false
        desc: |
          Sets the url to the EDAM ontology file.
```
And then we can fetch that on startup if it's not present.

**Comment by hexylena** [comment]
> It's just the unit tests that mock out a config object in various places. Want me to fix that ?

Sure, that'd be ace!

File: lib/galaxy/webapps/galaxy/config_schema.yml
Comment (hexylena) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: Ah, that's a nice suggestion, sure. sounds good.

Comment (hexylena) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: Just to confirm because the diff looks odd: swap path for url, right? remove path completely?

**Comment by mvdbeek** [comment]
> * we have a copy of the EDAM ontology in the codebase. Surely there's a better way to handle this?

Downloading this file now from beta_edam_ontology_url. Can you rebase out config/EDAM_1.23.tsv  ?

I just gave it a go, it looks nice!

File: lib/galaxy/webapps/galaxy/config_schema.yml
Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: I think we want to keep both. If the path doesn't exist we download and store the file at the path.

Comment (nsoranzo) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: Why do we want the URL to be configurable? Can't we just download the latest version from http://edamontology.org/EDAM.tsv ?

Comment (hexylena) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: In case EDAM releases a new ontology and people want to update before we update galaxy?

Comment (nsoranzo) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: That URL is supposed to always point to the latest EDAM release.

Comment (hexylena) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: Oh, sorry, misread the URL. Then I guess we need to worry about caching a bit, or? People will have the edam from whenever they first try this setting, and it won't be updated?

Comment (mvdbeek) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: I thought there might be other places / custom ontologies people may want to load. If that's not true we can just make the URL a constant, sure.

Comment (nsoranzo) [reply lib/galaxy/webapps/galaxy/config_schema.yml]: > Oh, sorry, misread the URL. Then I guess we need to worry about caching a bit, or? People will have the edam from whenever they first try this setting, and it won't be updated?

Sure, we have a similar issue with the conda binary.

> I thought there might be other places / custom ontologies people may want to load. If that's not true we can just make the URL a constant, sure.

Given we are using the EDAM annotations in tools, I don't think we need to make that configurable at this stage.

**Comment by hexylena** [comment]
I've added section labels, and ordered all of the contents according to the alphabetically sorted. I thought it would be better, but, not sure how I feel about it in the end. 

image

**Comment by mvdbeek** [comment]
I like that, it is a consistent way to read the panel.

**Comment by hexylena** [comment]
Hid labels without children, ok, I definitely like that too.

image

File: lib/galaxy/tools/toolbox/base.py
Comment (jmchilton) [inline lib/galaxy/tools/toolbox/base.py:1024]: Can we instead make this an API parameter and have this more of a dynamic view of the toolbox?

Comment (hexylena) [reply lib/galaxy/tools/toolbox/base.py:1024]: That definitely sounds like a better API. Would you have any suggestions for that refactoring? (It's been a while since I've done dev type things) Doing both the edam version and non-edam version parsing, and building two internal toolboxes?

Comment (hexylena) [reply lib/galaxy/tools/toolbox/base.py:1024]: Just following up here @jmchilton in case you had any opinions on the implementation :)

**Comment by dannon** [comment]
I haven't tried this yet, but it aligns a lot with what @juleengraham has been working towards to support multiple user/admin/upstream defined toolsets in the tool panel.  Have you followed that work?  (I'm also out of date on the status of that, and haven't tried this yet, but I'm trying to catch up and wanted to comment before I forget!)

**Comment by hexylena** [comment]
> Have you followed that work?

That sounds fantastic, I have not heard about it, or followed it at all. Is there a tracking issue somewhere or so?

**Comment by juleengraham** [comment]
@hexylena Dannon is referring to PR #9996. I actually brought up this EDAM work in our JHU meeting last week and briefly started a discussion about how the two might be related. At the time we weren't really sure what EDAM tags were. I'm sure we're going to talk about it more later today. I know in your original issue, you mention admins wanting to have their own lists and I think that would be a great way to connect these two projects.

Correct me if I'm wrong but EDAM tags seem to be a part of a tool's configuration (so specified by the tool's creator/maintainer?), but more standardised because EDAM comes from a biological background/definition.

**Comment by hexylena** [comment]
@juleengraham ah neat, I hadn't seen that. Admins having their own organisations of the tool panel was definitely discussed in the original, but I think it was potentially a bit orthogonal to the goals of an EDAM view? (i.e. I didn't have bandwidth to tackle that problem.)

> Correct me if I'm wrong but EDAM tags seem to be a part of a tool's configuration (so specified by the tool's creator/maintainer?), but more standardised because EDAM comes from a biological background/definition.

This is exactly correct. Tool developers tag their tools with the functions and topics of the tool. It's not an admin or user controlled feature. But it could definitely fit into the "one of many" views of a tool panel, like it seems you've been tasked with!

**Comment by juleengraham** [comment]
>> This is exactly correct. Tool developers tag their tools with the functions and topics of the tool. It's not an admin or user controlled feature. But it could definitely fit into the "one of many" views of a tool panel, like it seems you've been tasked with!

@hexylena noted! The toolsets weren't necessarily intended to be based on EDAM tags but I could see that being a use case. I started out working based on the EU's pre-curated sets (https://github.com/usegalaxy-eu/usegalaxy-eu-tools) but I've changed the implementation a bit in hopes that it'll be more user-friendly for the admins. I think for now, the lists were meant to be created by the admin of the instance and no individual users

**Comment by mvdbeek** [comment]
As an optional / dev setup this looks fine to me and certainly is a great incentive for more EDAM. Would you be opposed to merging as is (after resolving conflicts) @jmchilton?

**Review by mvdbeek — APPROVED** [review]
Can you resolve conflicts @hexylena ?

**Comment by hexylena** [comment]
Yes, absolutely @mvdbeek

**Comment by hexylena** [comment]
Resolved.

**Review by jmchilton — APPROVED** [review]
Approve after linting and tests are passed. Sorry for dropping the ball on the review of this. I'd like to take a pass at making it a bit more dynamic but this is a cool step forward. Thanks so much!

**Comment by hexylena** [comment]
No worries. would love for it to be more dynamic too. I look forward to our
multi view toolbox future!

John Chilton <notifications@github.com> schrieb am Do., 7. Jan. 2021, 23:20:

**Comment by mvdbeek** [comment]
@ic4f I'm sure this is an easy one for you, what did we need to do for this unit test failure?
```
____________ test_config_defaults[beta_edam_toolbox_ontology_path] _____________

test_data = TestData(key='beta_edam_toolbox_ontology_path', expected='EDAM.tsv', loaded='/home/circleci/repo/database/EDAM.tsv')

    @pytest.mark.parametrize('test_data', get_config_data(), ids=get_key)
    def test_config_defaults(test_data):
>       assert test_data.expected == test_data.loaded
E       AssertionError: assert 'EDAM.tsv' == '/home/circle...base/EDAM.tsv'
E         - /home/circleci/repo/database/EDAM.tsv
E         + EDAM.tsv

test/unit/config/test_config_values.py:227: AssertionError
```

**Comment by ic4f** [comment]
I'm looking into this - give me a few minutes :-)

**Comment by ic4f** [comment]
@hexylena You need to add the following to this dict: https://github.com/galaxyproject/galaxy/blob/release_20.09/test/unit/config/test_config_values.py#L100
```
'beta_edam_toolbox_ontology_path': self._in_data_dir('EDAM.tsv'), 
```
It tells the test how to prepare the expected value: it places the 'EDAM.tsv' file inside the `data_dir` (which is created at test runtime).

Sorry for the lacking docs (it's all in the comments in that test file).

**Comment by dannon** [comment]
@mvdbeek Yeah, no opposition here.

**Comment by mvdbeek** [comment]
> Sorry for the lacking docs (it's all in the comments in that test file).

The test could maybe tell us that, so if it fails we know what to do ?

**Comment by ic4f** [comment]
@mvdbeek yes (a message can be passed to the test with the rest of the data for cases where such a message makes sense). However, the path generation for the test can be automated: we have the key, we have the schema which contains the `path_resolves_to` value. I think that would be the better way to do it: then we don't have to require this extra step from anyone who's editing a path option in the schema. (writing accompanying tests is good; having to modify a nontrivial testing module for trivial things like a new config option - less so, I think).

**Pull request merged**
